src/combinator.rs

```diff
@@ -526,6 +526,80 @@ where
     go_extra!(I::Span);
 }
 
+
+/// See [`Parser::try_foldl`].
+pub struct TryFoldl<F, A, B, OB, E> {
+    pub(crate) parser_a: A,
+    pub(crate) parser_b: B,
+    pub(crate) folder: F,
+    #[cfg(debug_assertions)]
+    pub(crate) location: Location<'static>,
+    #[allow(dead_code)]
+    pub(crate) phantom: EmptyPhantom<(OB, E)>,
+}
+
+impl<F: Copy, A: Copy, B: Copy, OB, E> Copy for TryFoldl<F, A, B, OB, E> {}
+impl<F: Clone, A: Clone, B: Clone, OB, E> Clone for TryFoldl<F, A, B, OB, E> {
+    fn clone(&self) -> Self {
+        Self {
+            parser_a: self.parser_a.clone(),
+            parser_b: self.parser_b.clone(),
+            folder: self.folder.clone(),
+            #[cfg(debug_assertions)]
+            location: self.location,
+            phantom: EmptyPhantom::new(),
+        }
+    }
+}
+
+impl<'src, I, F, A, B, OA, OB, E> Parser<'src, I, OA, E> for TryFoldl<F, A, B, OB, E>
+where
+    I: Input<'src>,
+    A: Parser<'src, I, OA, E>,
+    B: IterParser<'src, I, OB, E>,
+    E: ParserExtra<'src, I>,
+    F: Fn(OA, OB, I::Span) -> Result<OA, E::Error>,
```

Comment:
`I::Span` shouldn't be passed in here. If you need a span, it's better to also create a `TryFoldlWith` combinator. Like the other `XWith` combinators, it generalises to accessing spans, context, and state.

Additionally, there's already a clue in `FoldlWith` regarding the span: the span starts from before everything (`before_all`).

Reply:
I partly based this of `try_map` and `foldl`. In `try_map` we do get a span, but of course I can split that to a `TryFoldlWith` and do it similar to `foldl_with` if you prefer.

Reply:
Hmm, I do understand this reasoning on a second pass. It's not possible to generate an error without having access to the span anyway. Perhaps `try_foldl` should just include the extra parameter (i.e: like `foldl_with`) by default?

Reply:
I'm not sure. I like the consistency with `try_map`, but I'm not sure if its worth to implement a second entire `try_foldl_with`. I've taken the easy route for now.